core/src/execution/datafusion/planner.rs

```diff
@@ -1039,6 +1039,21 @@ impl PhysicalPlanner {
                 .collect();
             let full_schema = Arc::new(Schema::new(all_fields));
 
+            // Because we cast dictionary array to array in scan operator,
+            // we need to change dictionary type to data type for join filter expression.
+            let fields: Vec<_> = full_schema
+                .fields()
+                .iter()
+                .map(|f| match f.data_type() {
+                    DataType::Dictionary(_, val_type) => {
+                        Arc::new(Field::new(f.name(), val_type.as_ref().clone(), true))
```

Comment:
Should we preserve the fields nullable flag here?

```suggestion
                        Arc::new(Field::new(f.name(), val_type.as_ref().clone(), f.is_nullable()))
```

Reply:
Yea